Simplify Auto Bed Leveling Positions Configuration

tkoopman opened the pull request:
Calculates the ABL Grid boundaries based on bed size & probe offsets.

tkoopman committed:
Simplify Auto Bed Leveling Positions Configuration

Calculates the ABL Grid boundaries based on bed size & probe offsets.

FHeilmann commented on Marlin/Configuration.h:
```diff
@@ -808,10 +808,21 @@
+  #else
+    #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER + PROBE_BED_MARGIN)
+    #define BACK_PROBE_BED_POSITION (Y_MAX_POS - PROBE_BED_MARGIN)
+  #endif
```
You don't need two separaget #if #else #endif blocks if they both theck for the same thing. You can consolidate your code into one of these blocks.

tkoopman replied:
Copy and Paste + Testing Fail. My bad :)
Tested with probe offsets both + and both -, so missed typo for when they different.
Fixed.

Blue-Marlin replied:
I'm quite sure the second
```
#if X_PROBE_OFFSET_FROM_EXTRUDER < 0
```
should be
```
#if Y_PROBE_OFFSET_FROM_EXTRUDER < 0
```

fiveangle replied:
I've had mine this for a while now:

```
  // Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

Submitting this simple PR in order to get my github feet wet has been on my todo list for quite some time and haven't gotten to it, so please take if you think this works better.

tkoopman replied:
Looks good to me. Inline If statements neater.

los7702 replied:
Just one question? If zero is in the middle then left pos could be min probe edge to the rigt of middle...??

los7702 replied:
This is my suggestion...
```cpp
// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
#define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
#define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
#define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
#define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

thinkyhead replied:
Generally all good suggestions. I think the auto-assigning of these values should be moved into `Conditionals_POST.h`, in this manner:
```cpp
#ifndef LEFT_PROBE_BED_POSITION
  #define LEFT_PROBE_BED_POSITION  ((X_PROBE_OFFSET_FROM_EXTRUDER) <= (MIN_PROBE_EDGE) ? (X_MIN_POS + MIN_PROBE_EDGE) : (X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER))
#endif
```
...and so on. (Note the extra parentheses!) Then, comment out the lines in `Configuration.h` and add documentation that these values are automatically set and only need to be changed under special conditions. Some of the example configurations might provide overrides. Most of them will probably be able to leave them undefined.

los7702 replied:
@thinkyhead that is even better. if not defined, then calculate...
1 vote from me

fiveangle replied:
@los7702, yeah... wasn't thinking of center-origined 'bots.  Automating defaults outside of Config is perfect.

tkoopman committed:
Simplify Auto Bed Leveling Positions Configuration

Typo Fixed

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -808,10 +808,21 @@
+  #else
+    #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER + PROBE_BED_MARGIN)
+    #define BACK_PROBE_BED_POSITION (Y_MAX_POS - PROBE_BED_MARGIN)
+  #endif
```
I've had mine this for a while now:

```
// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? MIN_PROBE_EDGE : Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
```

fiveangle replied:
Was going to make this my first PR to get my feet wet, but you beat me to it, so please take if you think this works better.

los7702 replied:
This is my suggestion...
`// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)`

los7702 replied:
This is my suggestion...
`// Set the boundaries for probing (where the probe can reach). Calculations set to max dimensions inset by MIN_PROBE_EDGE. Replace calculations below with hard-coded values to set manually.
  #define LEFT_PROBE_BED_POSITION  (X_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? X_MIN_POS + MIN_PROBE_EDGE : X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER) 
  #define RIGHT_PROBE_BED_POSITION (X_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? X_MAX_POS - MIN_PROBE_EDGE : X_MAX_POS + X_PROBE_OFFSET_FROM_EXTRUDER)
  #define FRONT_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER <= MIN_PROBE_EDGE ? Y_MIN_POS + MIN_PROBE_EDGE : Y_MIN_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)
  #define BACK_PROBE_BED_POSITION (Y_PROBE_OFFSET_FROM_EXTRUDER >= -MIN_PROBE_EDGE ? Y_MAX_POS - MIN_PROBE_EDGE : Y_MAX_POS + Y_PROBE_OFFSET_FROM_EXTRUDER)`

tkoopman committed:
Revert "Simplify Auto Bed Leveling Positions Configuration"

This reverts commit 4de5ef1baf207eecfabd3c635a3dc8d0695218eb.

tkoopman committed:
Revert "Simplify Auto Bed Leveling Positions Configuration"

This reverts commit 2746f35be8cc915d742bfc06eec6df19ace22b0f.

tkoopman committed:
Simplify Auto Bed Leveling Positions Configuration - Updated

Moved calculations to Conditionals_post.h and now just comment out
definitions in Configuration.h.

tkoopman commented:
Updated based on comments. Tested fine on my printer.
How's it look to you guys now?

Tim.

fiveangle commented:
Thinkyhead stated: "(Note the extra parentheses!)" but I don't see them in your code.  Probably a good idea to be doubly-sure.

fiveangle commented:
Nitpicking, but you could probably change:

// The Z probe minimum outer margin (to validate G29 parameters).

to:

// The Z probe minimum outer margin (to validate G29 parameters and auto-calculate default probe bed positions).

los7702 commented:
Great job.
Looks good to me.

thinkyhead commented:
Please only submit PRs to the `RCBugFix` branch.

thinkyhead commented:
A comprehensive patch should take account of the probe offsets, and not just the `MIN_PROBE_EDGE` (which is legacy cruft). We now require C++11, so sanity checking can use `static_assert`.
